### PR TITLE
feat: graph data structure.

### PR DESCRIPTION
**Pull request by fhlasek**
#### Description of Change
#1004 The data structure can be further enhanced by adding weighted edges or more functionality for various constructors or IO operations. Aiming for the very minimal usable skeleton to make it easier to review.

#### Checklist

- [X] Added description of change
- [X] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [X] Added tests and example, test must pass
- [X] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [X] Relevant documentation/comments is changed or added
- [X] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [X] Search previous suggestions before making a new one, as yours may be a duplicate.
- [X] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1065"></a>

### CODE REVIEW

**Comment by fhlasek** [comment]
@kvedala can you please help me here? It seems like the code checker is aiming to process the file with C language rather than C++
```
error: invalid argument '-std=c++11' not allowed with 'C' [clang-diagnostic-error]
/home/runner/work/C-Plus-Plus/C-Plus-Plus/data_structures/graph.h:21:10: error: 'cassert' file not found [clang-diagnostic-error]
#include <cassert>
         ^
```

**Comment by kvedala** [comment]
1. `.h` file is considered by `clang` as a C-file and not a CPP. Changing the `graph.h` to `graph.hpp` might work. 
2. the filenames are ambiguous and might be better to use `directed_graph.h` and `directed_graph.cpp`. But see point 3 below
3. `graph.h` is simple enough to be included in the cpp file itself. Hence, a single `directed_graph.cpp might suffice here?

**Comment by fhlasek** [comment]
@kvedala thanks for the suggestion, renaming to `.hpp` seems to work. I don't quite follow what the other recommendations are. Are you happy with how the code is organised right now or do you want me to change it somehow?

**Comment by kvedala** [comment]
> @kvedala thanks for the suggestion, renaming to `.hpp` seems to work. I don't quite follow what the other recommendations are. Are you happy with how the code is organised right now or do you want me to change it somehow?

I that a previous discussion led to us thinking that having a `graph.h` file generic to all might be a good way to go. However, thinking more about it, it might actually cause more issues to contributors. It is really difficult to maintain a generic `graph.h` code that would cater every contributor's needs and for the diverse graph algorithms there are but would not be able to readily use the header file.

**Comment by fhlasek** [comment]
@kvedala I suggest we check the code in, but don't enforce using this structure for every graph algorithm. I understand that certain algorithms might not be an easy fit for the simple structure representation, and they'll have a full discretion to define their own input structure. I believe this way we can greatly reduce the code duplication and make it actually much easier to maintain while keeping the flexibility. What do you think?

**Comment by fhlasek** [comment]
Alternatively, since the main purpose of this repo is educational, we can have a very minimalistic version of this common piece of code in form of a _oneliner_ typedef that'd define the graph structure as an adjacency list.

**Comment by fhlasek** [comment]
@kvedala do you think #1076 would be a better approach?

**Comment by kvedala** [comment]
Sure @fhlasek 
1. Including the in the repo does no harm and we can proceed with the merge.
2. Having a skeleton file as in this repo is fine but need not be enforced to be used by the contributors. 
3. The approach in #1076 is good and in accordance with modern C++ practices. Again, need not be enforced.

File: data_structures/test_graph.cpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
+#include <cassert>
+#include <set>
+
+#include "./graph.hpp"
+
+/** Test function */
```
Comment (kvedala) [inline data_structures/test_graph.cpp:6]:
```suggestion
/** Test function 
 * @returns void
 */
```

File: data_structures/graph.hpp [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+ */
+namespace data_structures {
+
+/**
+ * @brief Data structure storing a directed graph as an adjacency list.
+ */
+class directed_graph {
+ public:
+    /** @brief For every vertex a list of its neigbors is stored.  */
+    std::vector<std::list<int> > neighbors;
```
Comment (kvedala) [inline data_structures/graph.hpp:37]: should this be made private?

**Pull request closed without merging**